src/wasm/wasm-type.cpp

```diff
@@ -1319,85 +1333,102 @@ void Canonicalizer::scanType(Type* type) {
   }
 }
 
-void Canonicalizer::makeAncestorsFixedPoint() {
-  // Naively calculate the transitive closure of the reachability graph.
+void Canonicalizer::findSelfReferentialHeapTypes() {
+  // Calculate the fixed point of the reachability relation.
+  auto fixedPoint = children;
   bool changed;
   do {
     changed = false;
-    for (auto& entry : ancestors) {
+    for (auto& entry : fixedPoint) {
       auto& succs = entry.second;
-      std::unordered_set<TypeID> nextAncestors;
+      std::unordered_set<TypeID> newSuccs;
       for (auto& other : succs) {
-        auto& otherAncestors = ancestors[other];
-        nextAncestors.insert(otherAncestors.begin(), otherAncestors.end());
+        auto& otherSuccs = fixedPoint[other];
+        newSuccs.insert(otherSuccs.begin(), otherSuccs.end());
       }
       size_t oldSize = succs.size();
-      succs.insert(nextAncestors.begin(), nextAncestors.end());
+      succs.insert(newSuccs.begin(), newSuccs.end());
       if (succs.size() != oldSize) {
         changed = true;
       }
     }
   } while (changed);
+
+  // Find HeapTypes that reach themselves
+  for (auto& entry : builder.impl->entries) {
+    TypeID id = entry.get().getID();
+    auto it = fixedPoint.find(id);
+    if (it != fixedPoint.end() && it->second.count(id)) {
+      selfReferentialHeapTypes.insert(id);
+    }
+  }
 }
 
 std::vector<Canonicalizer::Item> Canonicalizer::getOrderedItems() {
-  // Topologically sort the Types and HeapTypes so that all children are
-  // canonicalized before their parents.
+  // In order to canonicalize Types and HeapTypes without leaking any temporary
+  // types into the global type store, we need to canonicalize children before
+  // parents. In the case of recursive types, this is not possible due to cycles
+  // in the parent-child relation. In principle that can be overcome by
+  // canonicalizing type structures rather than type contents, but for now we
+  // instead cut corners and break the cycles by skipping canonicalization of
+  // self-referential HeapTypes. This works because the path from any Type or
+  // HeapType to itself in the parent-child relation must go through some
+  // self-referential HeapType; it is not possible to have a cycle composed only
+  // of Types, for example. In effect this means that we have a nominal type
+  // system for self-referential HeapTypes and a structural type system for
+  // everything else. Eventually we will have to implement something more
+  // consistent, but this is good enough for getting prototype toolchains up and
+  // running.
+
+  // Remove self-referential HeapTypes to cut cycles.
+  auto childrenDAG = children;
+  for (TypeID id : selfReferentialHeapTypes) {
+    childrenDAG.erase(id);
+    for (auto& kv : childrenDAG) {
+      kv.second.erase(id);
+    }
+  }
```

Comment:
I don't know if this matters much, but for performance I suspect the reversed loops might be faster. I am assuming that most types are not self-referential - does that sound reasonable?

If so, then cache locality would benefit from having the inner loop be on the smaller thing, `selfReferentialHeapTypes`. But then you'd need two loops to avoid iterator invalidation I guess,
```cpp
for (auto& kv : childrenDAG) {
  kv.second.erase(selfReferentialHeapTypes.begin(), selfReferentialHeapTypes.end());
}
childrenDAG.erase(selfReferentialHeapTypes.begin(), selfReferentialHeapTypes.end());
```

Reply:
IIUC, the variant of the `erase` method that takes iterators only accepts iterators into the collection it is being called on. So I don't think it's valid to pass iterators of `selfReferentialHeapTypes` to the `erase` method of the sets in `childrenDAG`. I could still invert the loops, but then there would be more loops overall :( Overall I prefer the current version.

```
  for (TypeID id : selfReferentialHeapTypes) {
    childrenDAG.erase(id);
  }
  for (auto& kv : childrenDAG) {
    for (TypeID id : selfReferentialHeapTypes) {
      kv.second.erase(id);
    }
  }
```

Reply:
Oh right, I mixed up that API...

Given it's more code it's not worth doing this possible optimization now, I agree. Maybe a TODO to investigate? Then if we profile in the future and see this function is slow, we'd have a hint.